README.md
# dmsh
Dartmouth Menu Shell

```
dmsh(1) 			     MENU			       dmsh(1)

NAME
       dmsh - Dartmouth Menu Shell

SYNOPSIS
       dmsh [options] [menufile]

OPTIONS
       No options are supported at this time.

DESCRIPTION
       dmsh  is a curses-based menu generator.	It has the ability to run pro-
       grams and access additional menu  files.   The  default	menu  file  is
       called main.mnu in the current working directory.  A menu file pathname
       may also be given on the command line.

MENU FILE DESCRIPTION
       The menu file is a text	file  consisting  of  ":"  delimited  records.
       Lines  beginning  with a '#' are considered comments and ignored.  Each
       record (line) has three fields.	There is a line length	limit  of  512
       bytes

       For  backward  compatability,  the  first  record  in  the menu file is
       reserved for a title of the menu, and is always treated as a "T"  line.
       This  feature  may  go  away.  Also for backward compatability, a blank
       line in the menu file is treated as a "T" line consisting of  a	series
       of dashes, to make simple screen dividers for menu groups.

   Fields
       Field 1 is the description that will show up as a menu selection on the
       screen.	It has a length limit of 80 characters.

       Field 2 is a 1-character (case insensitive)  flag  indicating  how  the
       rest  of the record is to be interpreted.  It is selected from the fol-
       lowing list.  Any other characters in that field will cause the	record
       to be displayed on screen as "(invalid)".

       Field 3 -- the rest of the line, consists of the command to be executed
       when this menu item is selected.  It is passed to the system()  library
       function  and so can contain shell metacharacters.  In the case of sub-
       menus, field 3 is the pathname of the menu file to be opened.

   Field 2 Options
       T      "T" is used to indicate  that field 1 contains  text  to	become
	      part  of	the displayed menu.  It is used for headers or section
	      separators, and any field 3  is  ignored.   No  menu  letter  is
	      assigned	to this line.  Completely blank lines may be generated
	      with a null first field.

       P      "P" indicates that field 3 contains a program or script  to  run
	      (via system()) when this menu item is selected.  dmsh will pause
	      after running the program, until the [RETURN] key is pressed

       N      "N" indicates that field 3 contains a program or script  to  run
	      (via  system()) when this menu item is selected.	On completion,
	      dmsh continues immediately and redraws the current menu.	If the
	      system()	call  exits with a non-zero value, then a warning mes-
	      sage is displayed, and dmsh pauses to allow error messages to be
	      seen, as if "P" had been used.

       Q      "Q"  is interpreted as for "N", but dmsh exits immediately after
	      running the command.  The menu is not redisplayed.

       *      "*" is used  to  disable	menu  items.   The  description  still
	      appears,	but the string "(unavailable)" is appended.  It can be
	      used to disable an item without removing the item completely.

       M      "M" indicates a sub-menu.  The string "(menu)"  is  appended  to
	      the  description and field 3 is interpreted as a file name to be
	      openend and read as a new menu  file.   The  "~user"  syntax  is
	      interpreted if found at the start of the pathname.

   Field 3 options
       The  program  can be a series of shell commands, and programs delimited
       by the semi-colon.  Any /bin/sh metacharacters may be used.   The  com-
       mand  may  contain ":" since no further parsing for fields is performed
       after locating the start of field 3 on the line.

MENU SELECTION
       Each  menu selection is given a letter label,  starting with [a].

       The menu item [q] is added to every menu, and either quits the  program
       (if  at	the  tope level) or returns to the previous menu (if in a sub-
       menu).  The ESC key may also be used in place of 'q'.

       Pressing a single  letter  key  will  run  the  corresponding  command.
       Pressing  the uppercase menu letter will display what would be run, for
       debugging.

Examples
       Main Menu:T:noop
       Internet Tools:M:sub1.mnu
       Today's Date:p:clear;date
       Who's On:p:clear;finger|sort
       Compile a Program:*:noop
       Shell:N:clear;echo "Type exit to return";PS1='$ ';export PS1;$SHELL
       Directory Listing:p:clear;ls -lA
       <this empty line will show as dashes>

       If the program is invoked with the above file as the argument, the display
       appears as follows:
	  Dartmouth Menu Shell, version: <version>  <date and time>

	       Main Menu
	   [a]	Internet Tools (menu)
	   [b]	Today's Date
	   [c]	Who's On
	   [d]	Compile a Program (unavailable)
	   [e]	Shell
	   [f]	Directory Listing
	    ---------------
	   [q]	Quit

	 Select Value:

	MESSAGES:   Enter a letter or [q] to quit

FILES
       main.mnu

REVISIONS
       Sun Jan 17 07:50:19 CST 2021

Dartmouth College		   Jan 2021			       dmsh(1)
```


CHANGES
V 0.8 changes to accomodate Bio-Epi group use on Brooks and Cascade:

Redefined MESSAGE from 19 to 21 to match actual executable on Belknap
Redefined MAXLINE from 250 to 512 to accomodate enormous scripts in bioepi menus
(Also made to v1.x).  Increased the fixed size of the object field in the menu struct
to also be MAXLINE (was 256)


V 1.0 2001/06/01 RB

- Redefined MESSAGE from 21 to (LINES-5).  This is the screen position of the message
area at the bottom of the screen, and tying it to LINES makes it work on any size
screen supported by curses.  (Still no support for menus which exceed a single screen)
- Increased maximum size of desc field to 80 (was 50)
- Added 1-byte "label" field to the menu struct so that labels can be something other
than just index into the array of menu structs.  Experimented with using numbers instead
of letters, but this limits us at 9 items.
- label/separator lines do _not_ get a letter assigned to them, only active lines.
- miscellaneous code cleanup and buffer overrun protection.
- menu file syntax cleanup - explicit support for display labels ("T" in field 2)
- syntax to run a program and quit ("Q" in field 2)

V 1.01 2002/12/04 RB

- Redefined MAXLINE from 250 to 512 to accomodate enormous scripts in bioepi menus
(it would be nice to allocate this buffer dynamically and accomodate line continuation
with "\" escape, but that would be a larger change to the code).  MAXLINE is the
buffer used for the initial file parsing.  Increased the fixed size of the object field
in the menu struct to also be MAXLINE (was 256).
- Added support for resolving "~user" at the start of a pathname to a submenu file.
Variable newpath[MAXPATH] and some char * pointers added.  Uses getpwnam().
- Added support for a debug mode - selecting a menu item using the equivalent capital
letter just displays what the command would do, rather than execute the command (also
would be nice to have a debug option that only worked this way).
- Any unparseable item in field 2 is no longer assigned a letter code.  (Original
version defaults to assigning these)
- menu file syntax cleanup - explicitly skip over comment lines.
- Use bold to highlight sub-menus 
- Use bold for menu text lines - optionally
- Change error message for selecting an "unavailable" item - allow them to be selected
  so that they can be displayed in debug mode.

V 1.02 2003/05/16 RB

- Placed all source under RCS. Initial checkin is the 1.01 release.
- Extended "debug" mode so an uppercase sub-menu seletion displays the filename that would be
opened.

V 1.03 2006/04/12 RB

- Added function prototypes to remove compiler warnings, and to ensure no pointer-int assumptions.
  Code was segfaulting when compiled on 64-bit systems because of this
- Added check on return status of the system() call in runprog(), to allow user to see error
  messages and know the status was non-zero.
  Compiled and tested on AIX and Linux (32/64 bit)

V 1.04 2021/01/17 PS

- commented out malloc.h in menu.h as malloc is defined in stdlib.h now.  This enabled compiling on
  MacOS which doesn't have a malloc.h header
- change Makefile to use ncurses library 
- placed all source under GitHub.  Initial checkin is the 1.04 release.

IDEAS:
- Add color
- Have a configure script available
- Skip dashes on unparseable lines except in backward compatability mode ?
- Beep on errors ?
- optionally skip the main header (version/date)
- optionally skip interpreting the first line as a menu header - make it explicitly
  a T line
- need to sanity check the "todo" field and ensure it is only 1 character, and
  explicitly define the action if the value is unexpected.
- Add a -o option in invoke old mode and use that for all the backward compatability stuff.


main.c
/* Dmsh: Dartmouth Menu Shell */
/* $Header: /afs/@cell/uother/source/sep/dmsh/dmsh-1.0/RCS/main.c,v 1.3 2006/04/12 16:10:18 richard Exp $ */

/**********************************************************************
 
       Copyright (c) 1995 by the Trustees of Dartmouth College.

   Permission to use, copy, modify and distribute these programs and
   their  documentation  for  any  purpose and without fee is hereby
   granted, provided that this copyright notice and  permission  ap-
   pear on all copies and supporting documentation.  The name of the
   Trustees of Dartmouth College may not be used in  advertising  or
   publicity  relating  to  distribution  of  these programs without
   specific prior permission.  Neither  the  Trustees  of  Dartmouth
   College nor the author make representations about the suitability
   of this software for any purpose.  It is provided "as is" without
   expressed or implied warranty.

************************************************************************/

#include "menu.h"

int main(argc,argv)
int argc;
char *argv[];
{
    void die();
    char menufile[MAXLINE];

#ifdef LOGGING
    /* 
    Dartmouth Research Computing uses LOCAL2 for application usage logging, but
    this syntax does not match the standard form used by other wrappers. 
    LOCAL4 is unassigned as on 2006/04/11 - use that.  LOCAL0 is used by portsentry
    */
    openlog("admin",LOG_PID,LOG_LOCAL4);
    syslog(LOG_INFO,"MENU version %s Login Starting",VERSION);
#endif
    signal(SIGINT, (void *)die);

    if (argc < 2)
        strcpy(menufile,"main.mnu");
    else
        strcpy(menufile,argv[1]);

    initscr();
    initinput();
    wrap(menufile);
    die();
    return 0;
}
/******************************************************************************
 * INITINPUT - Set up curses
 *****************************************************************************/
void initinput()
{
    raw();
    echo();
    nonl();
}
/******************************************************************************
 * DIE - This function is where to end entire program
 *****************************************************************************/
void die()
{
    signal(SIGINT, SIG_IGN);
    mvcur(0, COLS-1, LINES-1, 0);
    endwin();
    exit(0);
}
/******************************************************************************
 * WRAP - This wraps up the whole program.
 *****************************************************************************/
void wrap(menufile)
char menufile[];
{
    intro(VERSION);
    menu(menufile);
    return;
}
/******************************************************************************
 * INTRO - This prints the program name, version and current time at the top
 *         of the display.
 *****************************************************************************/
void intro(version)
char version[];
{
    int x1, x2, y;
    char *tp;

    /* maybe horizontally centre this in the window, after computing the length */
    
    move(1,2);
    standout();
    addstr(" Dartmouth Menu Shell, version: ");
    getyx(stdscr, y, x1);
    printw("%s", version);
    getyx(stdscr, y, x2);
    while (x2++ - x1 < 6)
        addch(' ');
    tp = (char *)get_time();
    stripcrlf(tp);
    addstr(tp);
    addstr(" ");
    standend();
    move(MESSAGE,1);
    addstr("MESSAGES:");
    clrtobot();
    move(0,0);
    refresh();
}


menu.c
/* Dmsh: Dartmouth Menu Shell */
/* $Header: /afs/@cell/uother/source/sep/dmsh/dmsh-1.0/RCS/menu.c,v 1.3 2006/04/12 16:47:37 richard Exp $ */

/**********************************************************************
 
       Copyright (c) 1995 by the Trustees of Dartmouth College.

   Permission to use, copy, modify and distribute these programs and
   their  documentation  for  any  purpose and without fee is hereby
   granted, provided that this copyright notice and  permission  ap-
   pear on all copies and supporting documentation.  The name of the
   Trustees of Dartmouth College may not be used in  advertising  or
   publicity  relating  to  distribution  of  these programs without
   specific prior permission.  Neither  the  Trustees  of  Dartmouth
   College nor the author make representations about the suitability
   of this software for any purpose.  It is provided "as is" without
   expressed or implied warranty.

************************************************************************/
#include "menu.h"

void runprog(char *program, int debug, int pause);


/*****************************************************************************/
void menu(file)
char file[];
{
    FILE *menufd;
    MENU *item;
    int recnum = 0, i;
    char line[MAXLINE];
    char command[MAXLINE];
    char *p, *q, *u, newpath[MAXPATH], uname[MAXUSERNAME];
    struct passwd *pwent;
    static int state = 0;   /* used to determine which is top menu */
    int choose();

    state++;

    /* Added 12/16/02 RB for Bio-Epi use primarily */
    /* Resolve any "~user" at the start of the menu file pathname */
    /* We could also do this is the parse() function when the pathname is first read */
    if (file[0] == '~') {
        if ((p=strchr(file,'/'))) {
            /* need buffer overflow checking here */
            u = uname;
            for (q=file+1; q<p; q++,u++) { *u = *q; }
            *u = '\0';
#ifdef DEBUG
            sprintf(command,"Resolving username %s",uname);
            prtmsg(command,0);
            sleep(2);
#endif
            if ((pwent = getpwnam(uname)) != (struct passwd *)NULL) {
                strcpy(newpath,pwent->pw_dir);
                strcat(newpath,"/");
                strcat(newpath,p+1);
#ifdef DEBUG
                sprintf(command,"Attempting to open menu %s",newpath);
                prtmsg(command,0);
                sleep(2);
#endif
            } else {
                sprintf(command,"No such user: %s in menu line: %s",uname,file);
                prtmsg(command,0);
                sleep(2);
                return;
            }

        } else {
                sprintf(command,"Bad pathname specified in menu line: %s",file);
                prtmsg(command,0);
                sleep(2);
                return;
        }   
    } else {
        strncpy(newpath, file, MAXPATH);
    }


    /* open the menu file if it exists */
    if ((menufd = fopen(newpath,"r")) == NULL) {
        sprintf(command,"Unable to open menu %s",newpath);
        prtmsg(command,0);
        sleep(2);
        return;
    }

    /* count number of records in the menu file */
    while(fgetline(menufd,line,MAXLINE) != 0)
        recnum++;
    rewind(menufd);

    /* get enough memory to hold file dynamically */
    /* This allocates memory for non-menu items too - would be better to use a linked list
       allocated line by line */
    item = (MENU *)malloc(recnum*sizeof(MENU));
    if ( item == (MENU *)NULL) {
        sprintf(command,"Unable to allocate memory for menu %s",newpath);
        prtmsg(command,0);
        sleep(2);
        return;
    }

    for(i = 0; i < recnum; i++)  /* read file into malloc'd memory */
    {
        /* if we want to support comment or continuation lines, fgetline() would be the place */
        fgetline(menufd,line,MAXLINE);
        stripcrlf(line);
        parse(line,item+i);  /* parse the fields into structure */
    }

    do {
        display(item,recnum,state);
    } while(choose(item,recnum));

    free(item);
    state--;
    fclose(menufd);

    return;
}
/*****************************************************************************/
void 
parse(line, item)
char line[];
MENU *item;
{
    char *loc;

    if((loc = strchr(line,':')) == NULL) {
        /* any line with no ":" is counted as a menu separator */
        strcpy(item->desc,"---------------");
        item->todo = 'E';
        strcpy(item->object,"");
        return;
    }
    
    *loc = '\0';
    /* need to avoid buffer overflows, and limit to COLS-10 (maybe less left margin?) */
    strcpy(item->desc,line);  /* get first string field */
    ++loc;
    item->todo = *loc;        /* get second char field */
    loc += 2;
    strcpy(item->object,loc); /* get third string field */
    item->label = 0;
}
/*****************************************************************************/
void 
display(item,num,state)
MENU *item;
int num;
int state;
{
    int i, j;
    int offset=0140;  /* 0140 for letters, 060 for numbers */

    clrworkarea();
    move(3,4);standout();
    /* The first line is taken to be a menu title, regardless of how it is parsed */
    printw("    %s    ",item->desc); /* menu title */
    standend();
 
    j=1;
    for(i = 1; i < num; i++) {
        move(i+4,4);
        if((item+i)->todo == '*') {
            /* should unavailable items get an index code ? */
            /* it would be nice to "gray" this out, but attribute A_DIM */
            /* does nothing on some systems and does reverse video on others, so skip */
            printw("[%c]  %s (unavailable)",  j+offset,(item+i)->desc);
            /* (item+i)->label is now assigned, so we may select this item and give a customized error */
            (item+i)->label =  j+offset;
            j++;
        } else if((item+i)->todo == 'M' || (item+i)->todo == 'm') {
            printw("[%c]  %s ",j+offset, (item+i)->desc);
            attron(A_BOLD);
            printw("(menu)");
            attroff(A_BOLD);
            (item+i)->label =  j+offset;
            j++;
        } else if((item+i)->todo == 'E') {
            printw("  %s   ","--------------");
        } else if((item+i)->todo == 'T' || (item+i)->todo == 't') {
            attron(A_BOLD);
            printw("%s",(item+i)->desc); /* additional text - no associated action */
            attroff(A_BOLD);
        } else {
            /* default action is to get an assigned label - haven't sanity checked these yet */
            printw("[%c]  %s",j+offset,(item+i)->desc);
            (item+i)->label = j+offset ;
            j++;
        }
    }
    if (state == 1) {
        mvaddstr(i+5,4, "[q]  Quit");
        prtmsg("Enter a letter or [ESC/q] to quit",0);
    } else {
        mvaddstr(i+5,4, "[ESC/q]  Return to previous menu");
        prtmsg("Enter a letter or [ESC/q] to return",0);
    }
    mvaddstr(i+6,2, "Select Value: ");                     
    move(i+6,16);                       
    refresh();
}
/*****************************************************************************/
int
choose(item,num)
MENU *item;
int num;
{
    int c,i,d=0;
    /* void runprog(); */

    c = getch();
    if (isupper(c)) {
       /* Upper case input means set debug flag for the corresponding lower case menu entry */
       /* lower case means we must check for 'q' not 'Q' */
       d = 1;
       c = tolower(c);
    }
    /* Find the matching entry, by examining the labels */
    /* Original code used i = c - 0140 as an index into the array */
    for (i=1;i<num;i++) { if ((item+i)->label == c) break; }

    if(i >= 1 && i < num) {
        switch((item+i)->todo) {
            case 'M':
            case 'm':
                if (d) {
                   /* debug - just display the menu file we would have read */ 
                   prtmsg("Selection would read menu file:",0);
                   prtmsg((item+i)->object,1);
                   sleep(2);
                   clrmsg(1);
                } else {
                   /* recursively call menu() with a new file */
                   menu((item+i)->object);
                } 
                break;
            case 'N':
            case 'n':
                runprog((item+i)->object,d,0);
                break;
            case 'P':
            case 'p':
                runprog((item+i)->object,d,1);
                break;
            case '*':
                if (d) {
                   runprog((item+i)->object,d,1);
                } else { 
                   prtmsg("This item has been disabled",0);
                   sleep(2);
                   clrmsg(0);
                }
                break;
            case 'Q':
            case 'q':
                runprog((item+i)->object,d,0);
                /* Q means run this item and then immediately Quit */
                /* maybe have an option to exec() instead.  Should we clear the screen */
                clear();
                refresh();
                return 0;
                break;
        }
    } else if ( c == 'q' || c == '\033' )
        /* problem here if we had enough items to assign the letter 'q' */
        return 0;
    else {
        prtmsg("Invalid Entry, try again!",0);
        sleep(2);
        clrmsg(0);
    }
    return 1;
}
/*****************************************************************************/
void
runprog(program, debug, pause)
char program[];
int debug, pause;
{
    int c, pstat;
    char pcopy[MAXLINE];
    char *s;

    clear();
    refresh();
    move(1,1);
    endwin();  /* kill curses */

    /* New - display mode, rather than execute */
    if (debug) {
        /* display on screen, but split into lines at ";" */
        strncpy(pcopy, program, MAXLINE);
        s = strtok(pcopy, ";");
        printf("%s\n",s);
        while ((s = strtok(NULL, ";"))) {
            printf("%s\n",s);
        }
    } else {
        pstat = system(program); /* run the program */
        switch(pstat) {
        case -1:  /* fork failed */
          pause = 1;
          printf("Unable to run program: \"%s\"\n", program);
          break;
        case 0:   /* child process exited with 0 status */
          break;
        default:  /* child process exited with non-zero status */
          pause = 1;
          printf("WARNING: sh -c \"%s\" exited with status %d\n", program, pstat);
          break;
        }
    }

    if(debug||pause) {
        printf("\n\nPress [RETURN] to continue: ");
        /* IRIX wants a flush here or the prompt isn't seen */
        fflush(stdout);
        c = getchar();
    }

    initscr();         /* restart curses */
    intro(VERSION);
    initinput();
}
/*****************************************************************************/


misc.c
/* Dmsh: Dartmouth Menu Shell */
/* $Header: /afs/@cell/uother/source/sep/dmsh/dmsh-1.0/RCS/misc.c,v 1.3 2006/04/12 16:48:25 richard Exp $ */

/**********************************************************************
 
       Copyright (c) 1995 by the Trustees of Dartmouth College.

   Permission to use, copy, modify and distribute these programs and
   their  documentation  for  any  purpose and without fee is hereby
   granted, provided that this copyright notice and  permission  ap-
   pear on all copies and supporting documentation.  The name of the
   Trustees of Dartmouth College may not be used in  advertising  or
   publicity  relating  to  distribution  of  these programs without
   specific prior permission.  Neither  the  Trustees  of  Dartmouth
   College nor the author make representations about the suitability
   of this software for any purpose.  It is provided "as is" without
   expressed or implied warranty.

************************************************************************/
#include "menu.h"

/******************************************************************************
 * CLRMSG - This clears the message line selected
 *****************************************************************************/
void clrmsg(int offset)
{
    move(MESSAGE+offset,10);
    clrtoeol();
    move(0,0);
    refresh();
}
/******************************************************************************
 * CLRWORKAREA - This clears the window between line 3 and MESSAGE line.
 *****************************************************************************/
void clrworkarea()
{
    int i;

    for(i = 3; i < MESSAGE; i++)
    {
        move(i,0);
        clrtoeol();
    }
    refresh();
}

void xclrworkarea()
{
    int i;

    for(i = 3; i < MESSAGE; i++)
    {
        move(i,5);
        clrtoeol();
    }
}
/******************************************************************************
 * PRTMSG - This will print highlighted messages from the MESSAGE line on down.
 *          The offset argument determines which line from the MESSAGE line to
 *          print the message.
 *****************************************************************************/
void prtmsg(message,offset)
char *message;
int offset;
{
    clrmsg(offset);
    move(MESSAGE+offset,12);
    standout();
    printw(" %s ",message);
    standend();
    fflush(stdout);
    refresh();
    return;
}
/******************************************************************************
 *  GET_TIME - returns today's date and time followed by a '\n'
 *****************************************************************************/
char *
get_time()
{
    extern char *ctime();
    static double Tstart;
    static struct timeval start;


    (void) gettimeofday(&start, (struct timezone *)NULL);
    Tstart = (double) start.tv_sec + ((double)start.tv_usec)/1000000.0;
    return(ctime(&start.tv_sec));
}
/******************************************************************************
 * GET_INPUT - This routine takes string input in curses and allows for
 *             backspaces that are interpreted correctly
 *****************************************************************************/
void get_input(string)
char string[];
{
    char c; 
    int i = 0, y, x, Y, X;

    noecho();
    getyx(stdscr,Y,X);
    while((c = getch()) != '\r')
    {
        if (c == '\010' || c == '\177')
        {
            i -= 1;
            if (i < 0) 
            {
                i = 0;
                move(Y,X);
            }
            else
            {
                getyx(stdscr,y,x);
                move(y,x-1);
                delch();
            }
            refresh();
        }
        else
        {
            addch(c);
            refresh();
            string[i++] = c;
        }
    }
    string[i] = '\0';
    return;
}
/******************************************************************************
 * STRIPCRLF - Strip CR-LF from inputted strings here.
 *****************************************************************************/
void
stripcrlf(s)
char s[];
{
    char *cp;

    cp = s + strlen(s) - 1;
    while (cp >= s && (*cp == '\n' || *cp == '\r'))
        *cp-- = '\0';
    return;
}
/****************************************************************************** 
 * FGETLINE - to get a line of text from stdin
 *            borrowed from page 118 of "C! Programming and Practices"
 *            by M.Tim Grady who borrowed it from the K&R white book
 *
 *            6/27/90: Added file i/o instead of just stdin and
 *                     renamed from getline to fgetline
 *
 * 2002/12/30 RB Skip line if the first character is a '#'
 * 
 *****************************************************************************/
int fgetline(fd, s, lim)
FILE *fd;
char s[];
int lim;
{
    int c = 0, i = 0;

    while ((c=fgetc(fd)) != EOF && c == '#') {
        /* skip the rest of this line */
        while ((c=fgetc(fd)) != EOF && c != '\n') continue;
    }
    if (c == EOF) {
        s[i] = '\0';
        return i;
    } else if (c == '\n') {
        s[i++] = c;
        return i;
    } else {
        /* store the first character then continue as before */
        s[i++] = c;
        lim--;
    }

    while (--lim > 0 && (c=fgetc(fd)) != EOF && c != '\n')
        s[i++] = c;
    
    if (c == '\n')
        s[i++] = c;

    s[i] = '\0';

    return(i);
}
/******************************************************************************
 * GET_HIDDEN - This routine takes string input in curses and allows for
 *              backspaces that are interpreted correctly.
 *              All input is hidden and '*' are output instead.
 *****************************************************************************/
void get_hidden(string)
char string[];
{
    int c;
    int i = 0, y, x, Y, X;

    noecho();
    getyx(stdscr,Y,X);
    while((c = getch()) != '\r')
    {
        if (c == '\010' || c == '\177')
        {
            i -= 1;
            if (i < 0)
            {
                i = 0;
                move(Y,X);
            }
            else
            {
                getyx(stdscr,y,x);
                move(y,x-1);
                echo();
                delch();
                noecho();
            }
            refresh();
        }
        else
        {
            addch('*');
            refresh();
            string[i++] = c;
        }
    }
    string[i] = '\0';
    echo();
    return;
}



menu.h
/* Dmsh: Dartmouth Menu Shell */
/* Updated: Sun Jan 17 07:18:35 CST 2021 */
#include <stdio.h>
#include <string.h>	   /* Most systems use string.h now */
/* #include <strings.h> */ /* Some systems may need strings.h instead */
#include <ctype.h>
#include <curses.h>
#include <signal.h>
#include <sys/file.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <syslog.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>       /* Some systems may need this for sleep() */
/* #include <malloc.h> */    /* included in stdlib.h now */
#include <pwd.h>

#define VERSION "1.04"
#define MAXLINE 512       /* 2002/12/05 RB Increased from 250 for Bio-Epi */
#define MESSAGE (LINES-5) /* was fixed to 19 originally */

#ifndef MAXPATH           /* some systems define this in system headers    */
#define MAXPATH	256       /* 2002/12/16 RB Added for menu filename munging */
#endif

#define MAXUSERNAME 12


typedef struct menu {
   char label;
   char desc[80];
   char todo;
   char object[MAXLINE];  /* changed from 256, to match input buffer length */
}MENU;

/* local function prototypes */

void wrap(char *menufile);
void initinput();
void intro(char *version);
void menu(char *file);
void runprog(char *program, int debug, int pause);
void parse(char *line, MENU *item);
void display(MENU *item, int num, int state);

void clrmsg(int offset);
void clrworkarea();
void xclrworkarea();
void prtmsg(char *message,int offset);
char *get_time();
void get_input(char *string);
void stripcrlf(char *s);
int fgetline(FILE *fd, char *s, int lim);
void get_hidden(char *string);
